Thermochimica API support for MQMQA

parikshitbajpai opened the pull request:
Closes #24552 

Requires https://github.com/ORNL-CEES/thermochimica/pull/142

parikshitbajpai committed:
Check if MQMQA phase before species check (#24552)

mtnwalker commented:
One comment: it would be great to query total moles of elements in MSFL and in ideal_gas in addition to end members or species. Also naming convention for chemical potential instead of cp (heat capacity) may be confusing for some users - could try new naming convention. Thanks!

parikshitbajpai committed:
Thermochimica variable checks in ChemicalCompositionAction (#24552)

Adds the checks for input parameters against Thermochimica system to `ChemicalCompositinAction`. After deleting the checks in `ThermochimicaNodalData`, this will allow us to create aux variables only for elements / phases / species in the Thermochimica system. The changes also allow us to provide an *ALL* option for the input parameters in which case the values obtained from Thermochimica will be used to create aux variables.

Need to write aux variables directly in UO, finish tokenizing.

parikshitbajpai committed:
Thermochimica API using only UO (#24552)

This PR moves all the Thermochimica variable outputs to the `ThermochimicaNodalData` UO and gets rid of the intermediate `ThermochimicaAux` kernel. It also adds the *ALL* option and moves the input parameter checking and UO setup to `ChemicalCompositionAction`

parikshitbajpai committed:
Finalize changes to Thermochimica API (#24552)

This completes the changes required for API update.
- Includes bug fix for `ALL` species option.
- Update Thermochimica submodule
- Update documentation
- Regold exodus files for current tests
- Add an extra test for MQMQA pairs being treated as species. Also add
  gold file `MSTDB_out.e`.

parikshitbajpai committed:
Update Thermochimica submodule (#24552)

parikshitbajpai committed:
Tests fnd bug fixes for Thermochimica API (#2455)

Includes fix for failing tests due to `Thermochimica not defined` and a typo fix for `_vapor_pressures` being used instead of `_element_phases` in `ChemicalCompositionAction.C:L471`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1637421/) on b491632 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24553/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1637462/) on b491632 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>2d4e2b</th>
      <th colspan=4>#24553 b49163</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.56%</td>
      <td>85.56%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89164</td>
      <td>89161</td>
      <td>-3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15050</td>
      <td>15053</td>
      <td>+3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24553/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24553/coverage/framework)


## Modules coverage

### Chemical reactions

<table>
  <thead>
    <tr>
      <th></th>
      <th>2d4e2b</th>
      <th colspan=4>#24553 b49163</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>92.10%</td>
      <td>88.55%</td>
      <td>-3.56%</td>
      <td>76.29%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1178</td>
      <td>1237</td>
      <td>+59</td>
      <td>222</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>101</td>
      <td>160</td>
      <td>+59</td>
      <td>69</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24553/coverage_diff/chemical_reactions)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24553/coverage/chemical_reactions)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24553/coverage/xfem)

</details>

## Warnings

- `chemical_reactions` new line coverage rate 76.29% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1637462/) on b491632 wanted to post the following:


The following coverage requirement(s) failed:
- `chemical_reactions` coverage rate 88.55% is less than the required 92.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1639894/) on 62d3d70 wanted to post the following:


The following coverage requirement(s) failed:
- `chemical_reactions` coverage rate 91.20% is less than the required 92.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1640882/) on e58d49c wanted to post the following:


The following coverage requirement(s) failed:
- `chemical_reactions` coverage rate 91.62% is less than the required 92.0%

parikshitbajpai committed:
Ifndef directive for some Thermochimica errors (#24552)

Several mooseErrors in `ThermochimicaNodalData` are to safegaurd against possible deprecations / regressions in Thermochimica or `ChemicalCompositionAction`. They cannot be tested currently because variables are created automatically by `ChemicalCompositionAction` and possible errors are caught at that point. To hit the errors in ThermochimicaNodalData would require internal Thermochimica errors which we cannot "create" right now. To avoid failing code coverage due to this, enclosing these tests in `#ifndef NDEBUG`

dschwen commented on modules/chemical_reactions/doc/content/source/actions/ChemicalCompositionAction.md:
```diff
@@ -7,50 +7,58 @@ This action is designed for use with the thermochemistry library Thermochimica [
 The `ChemicalComposition` action is used to initiate the framework for thermochemical calculations using the thermodynamics solver [_Thermochimica_](https://github.com/ORNL-CEES/thermochimica). The action creates variables needed for the analysis, reads the thermodynamic model for material system, and sets the units for temperature, pressure, and elemental amounts. It can optionally initiate the  ...
 
-The `ChemicalComposition` action is also used to specify lists of chemical phases and chemical species for which concentrations should be output at each timestep. The phase and species names must match those in the data file specified. The `ChemicalComposition` action creates all AuxVariables and AuxKernels needed for output of phase concentration data. The list of phases is specified in the `out ...
+The `ChemicalComposition` action is also used to specify lists of chemical phases and chemical species for which concentrations should be output at each timestep. The phase and species names must match those in the data file specified. The `ChemicalComposition` action creates all AuxVariables needed for output of phase concentration data. The list of phases is specified in the `output_phases` var ...
```
```suggestion
The `ChemicalComposition` action is also used to specify lists of chemical phases and chemical species for which concentrations should be output at each timestep. The phase and species names must match those in the data file specified. The `ChemicalComposition` action creates all AuxVariables needed for output of phase concentration data, chemical potentials, vapor pressures, etc.. The list of phases is specified in the `output_phases` variable, and the list of species is specified in `output_species`. The species must be specified in the convention `phase_name:species_name` so that the desired species in the desired phase is uniquely identified. Alternatively, users can specify `ALL` to output all species. Other desired outputs such as element potentials, vapor pressures of species in gas phase and moles of elements in specified phases can be specified using `output_element_potentials`, `output_vapor_pressures`, and `output_element_phases` respectively. The `ChemicalComposition` action creates a UserObject derived from `ThermochimicaNodalData` which runs Thermochimica at each node and stores the output in the created auxvariables.
```

dschwen commented on modules/chemical_reactions/doc/content/source/actions/ChemicalCompositionAction.md:
```diff
@@ -7,50 +7,58 @@ This action is designed for use with the thermochemistry library Thermochimica [
 Chemical phase and species variables, as well as chemical potentials of elements, to be used and initiated in the model using in `ChemicalComposition` action are specified using the block `[GlobalParams]`, and vector `output_phases` as in [phaseact:globpar]:
 
-!listing id=phaseact:globpar caption=Chemical phases species to be used in the model, and list of output phases for which concentration data is desired.
+!listing id=phaseact:globpar caption=Chemical phases species to be used in the model, and list of desired output variables.
```
```suggestion
!listing id=phaseact:globpar start=GlobalParams end=[] include-end=True caption=Chemical phases species to be used in the model, and list of desired output variables.
```

dschwen commented on modules/chemical_reactions/include/actions/ChemicalCompositionAction.h:
```diff
@@ -21,6 +21,26 @@ class ChemicalCompositionAction : public Action
+
+  const std::vector<std::pair<std::string, std::string>> & speciesPhasePairs() const
+  {
+    return _token_species;
```
```suggestion
    return _tokenized_species;
```

dschwen commented on modules/chemical_reactions/include/actions/ChemicalCompositionAction.h:
```diff
@@ -52,4 +72,22 @@ class ChemicalCompositionAction : public Action
+  /// Name of the ThermochimicaNodalData UO to be set up
+  std::string _uo_name;
+
+  /// Mapped element IDs
```
```suggestion
  /// Atomic numbers of the selected elements
```

dschwen commented on modules/chemical_reactions/src/actions/ChemicalCompositionAction.C:
```diff
@@ -13,6 +13,7 @@
 #include "MooseMesh.h"
 #include "MooseUtils.h"
 #include "MooseUtils.h"
+#include <iostream>
```
```suggestion
```

dschwen commented on modules/chemical_reactions/src/actions/ChemicalCompositionAction.C:
```diff
@@ -68,12 +83,302 @@ ChemicalCompositionAction::ChemicalCompositionAction(const InputParameters & par
+  if (idbg != 0)
+    paramError("munit", "Cannot set mass unit in Thermochimica", idbg);
+
+  if (isParamValid("elements"))
```
isParamValid check is not necessary because the parameter is required

parikshitbajpai committed:
Address @dschwen PR comments (#224552)

Add FeTiVO gold and data files. Remove valid param test for elements.

dschwen commented on modules/chemical_reactions/doc/content/source/actions/ChemicalCompositionAction.md:
```diff
@@ -7,50 +7,53 @@ This action is designed for use with the thermochemistry library Thermochimica [
 
 The `ChemicalComposition` action is used to initiate the framework for thermochemical calculations using the thermodynamics solver [_Thermochimica_](https://github.com/ORNL-CEES/thermochimica). The action creates variables needed for the analysis, reads the thermodynamic model for material system, and sets the units for temperature, pressure, and elemental amounts. It can optionally initiate the  ...
```
```suggestion
The `ChemicalComposition` action is used to initiate the framework for thermochemical calculations using the thermodynamics solver [*Thermochimica*](https://github.com/ORNL-CEES/thermochimica). The action creates variables needed for the analysis, reads the thermodynamic model for material system, and sets the units for temperature, pressure, and elemental amounts. It can optionally initiate the amounts of elements used in the model.
```

dschwen commented on modules/chemical_reactions/doc/content/source/actions/ChemicalCompositionAction.md:
```diff
@@ -7,50 +7,53 @@ This action is designed for use with the thermochemistry library Thermochimica [
 The `ChemicalComposition` action is used to initiate the framework for thermochemical calculations using the thermodynamics solver [_Thermochimica_](https://github.com/ORNL-CEES/thermochimica). The action creates variables needed for the analysis, reads the thermodynamic model for material system, and sets the units for temperature, pressure, and elemental amounts. It can optionally initiate the  ...
 
-The `ChemicalComposition` action is also used to specify lists of chemical phases and chemical species for which concentrations should be output at each timestep. The phase and species names must match those in the data file specified. The `ChemicalComposition` action creates all AuxVariables and AuxKernels needed for output of phase concentration data. The list of phases is specified in the `out ...
+The `ChemicalComposition` action is also used to specify lists of chemical phases and chemical species for which concentrations should be output at each timestep. The phase and species names must match those in the data file specified. The `ChemicalComposition` action creates all AuxVariables needed for output of phase concentration data, chemical potentials, vapor pressures, etc. The list of pha ...
```
ThermochimicaNodalData should link to ThermochimicaNodalData.md

parikshitbajpai committed:
Address more comments from @dschwen (#224552)

parikshitbajpai committed:
Fix uo_name param in FeTiVO.i test (#24552)

parikshitbajpai committed:
Fix output species indexing when using ALL (#24552)

parikshitbajpai committed:
Improved performance and additional output species support (#24552)

- Reinitialization now supports `time`, `nodal` and `none` options. The default `nodal` option allows Thermochimica to use its internal reinitialization methods improving performance.
- The amount of output species can now be written in either `moles` or `mole_fraction`.
- Also restructures the steps for writing out moles of phase and chemical potentials to Aux variables. No more use of reinitialization data structs to write variables even when no/default reinitialization is used.

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 65977abfe982c9b00a93eaa52e47c169d9c28085
+Subproject commit 5252bb2d381f1d8ebcfdc650ac07928d3de69ad2
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1659973/) on 4a60d8d wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1679687/) on 45d06a3 wanted to post the following:


The following coverage requirement(s) failed:
- `chemical_reactions` coverage rate 91.94% is less than the required 92.0%

parikshitbajpai committed:
Add tests to improve coverage (#24552)

parikshitbajpai committed:
Fix test documentation and change parameters for FeTiVO CSV test (#24552)

dschwen committed:
Fix distributed tests

dschwen commented:
5a7d53a fixes distributed meshes and recover

dschwen reviewed (approved):
Looks good to me!

lindsayad commented:
This is causing a valgrind failure

lindsayad commented:
Error is due to accessing out of bounds in `moles_phase`

lindsayad commented:
Index is 0, so you are trying to access the `vector` at the `-1` element. Is the subtraction by -1 in the below erroneous?
```c++
         value = moles_phase[index - 1] * fraction;
```

parikshitbajpai commented:
> Index is 0, so you are trying to access the `vector` at the `-1` element. Is the subtraction by -1 in the below erroneous?
> 
> ```c++
>          value = moles_phase[index - 1] * fraction;
> ```

Checking now. I was juggling between C++ & Fortran and messed up. Will add a conditional there.

lindsayad commented:
Any luck?
